src/main/java/org/threadly/concurrent/future/Watchdog.java

```diff
@@ -0,0 +1,195 @@
+   * (which is lazily constructed).  This should be fine in most cases, but you can provide your 
+   * own scheduler if you want to avoid the thread creation (which is shared among all instances 
+   * that were constructed with this constructor).
+   * 
+   * @param timeoutInMillis Time in milliseconds that futures will be set to error if they are not done
+   * @param sendInterruptToTrackedThreads If {@code true}, and a thread is provided with the future, 
+   *                                        an interrupt will be sent on timeout
+   */
+  public Watchdog(long timeoutInMillis, boolean sendInterruptToTrackedThreads) {
+    this(getStaticScheduler(), timeoutInMillis, sendInterruptToTrackedThreads);
+  }
+  
+  /**
+   * Constructs a new {@link Watchdog} with a scheduler of your choosing.  It is critical that 
+   * this scheduler has a free thread available to inspect futures which may not have completed in 
+   * the given timeout.  You may want to use a limiter to ensure that there are threads available.
+   * 
+   * @param scheduler Scheduler to schedule task to look for expired futures
+   * @param timeoutInMillis Time in milliseconds that futures will be set to error if they are not done
+   * @param sendInterruptToTrackedThreads If {@code true}, and a thread is provided with the future, 
+   *                                        an interrupt will be sent on timeout
+   */
+  public Watchdog(SimpleSchedulerInterface scheduler, long timeoutInMillis, 
+                  boolean sendInterruptToTrackedThreads) {
+    this.scheduler = scheduler;
+    this.timeoutInMillis = timeoutInMillis;
+    this.sendInterruptToTrackedThreads = sendInterruptToTrackedThreads;
+    this.checkRunner = new CheckRunner();
+    this.futures = new ConcurrentLinkedQueue<FutureWrapper>();
+    this.checkRunnerStatus = new AtomicInteger(-1);
+  }
+  
+  /**
+   * Watch a given {@link ListenableFuture} to ensure that it completes within the constructed 
+   * time limit.  If the future is not marked as done by the time limit then it will be 
+   * completed by invoking {@link ListenableFuture#cancel(boolean)} with {@code true} to interrupt 
+   * the running thread.
+   * 
+   * @param future Future to inspect to ensure completion
+   */
+  public void watch(ListenableFuture<?> future) {
+    if (future == null || future.isDone()) {
+      return;
+    }
+    
+    final FutureWrapper fw = new FutureWrapper(future);
+    futures.add(fw);
+    // we attempt to remove the future on completion to reduce inspection needed
+    future.addListener(new Runnable() {
+      @Override
+      public void run() {
+        futures.remove(fw);
+      }
+    });
+    
+    while (true) {
+      if (checkRunnerStatus.get() == -1) {
+        if (checkRunnerStatus.compareAndSet(-1, 0)) {
+          scheduler.schedule(checkRunner, timeoutInMillis);
+          return;
+        }
+      } else if (checkRunnerStatus.get() == 1) {
```

Comment:
This is actually not true (at least for my processor).  Java does not do any .get() check internally, it attempts to send a CAS operation directly.  Based off my benchmarks doing this is NOT as efficent as doing the check first and only sending the CAS if you think it is likely to work.

I suppose the only time it might make sense to do a CAS every time is if the value was changing so much that it is just as likely to work, regardless of what the checked value was.